cache/cache.go

```diff
@@ -0,0 +1,496 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"cmp"
+	"slices"
+	"sort"
+	"sync"
+	"time"
+)
+
+// noExpiration is a sentinel value used to indicate no expiration time.
+// It is used instead of 0, to be able to sort items by expiration time ascending.
+const noExpiration = time.Second * 86400 * 365 * 10 // 10 years
+
+// Cache[T] is a thread-safe in-memory key/object store.
+// It can be used to store objects with optional expiration.
+type Cache[T any] struct {
+	*cache[T]
+	// keyFunc is used to make the key for objects stored in and retrieved from index, and
+	// should be deterministic.
+	keyFunc KeyFunc[T]
+}
+
+// item is an item stored in the cache.
+type item[T any] struct {
+	key string
+	// object is the item's object.
+	object T
+	// expiration is the item's expiration time.
+	expiration int64
+}
+
+type cache[T any] struct {
+	// index holds the cache index.
+	index map[string]*item[T]
+	// items is the store of elements in the cache.
+	items []*item[T]
+	// sorted indicates whether the items are sorted by expiration time.
+	// It is initially true, and set to false when the items are not sorted.
+	sorted bool
+	// capacity is the maximum number of index the cache can hold.
+	capacity   int
+	metrics    *cacheMetrics
+	labelsFunc GetLvsFunc[T]
+	janitor    *janitor[T]
+	closed     bool
+
+	mu sync.RWMutex
+}
+
+var _ Expirable[any] = &Cache[any]{}
+
+// New creates a new cache with the given configuration.
+func New[T any](capacity int, keyFunc KeyFunc[T], opts ...Options[T]) (*Cache[T], error) {
+	opt := storeOptions[T]{}
+	for _, o := range opts {
+		err := o(&opt)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if opt.registerer == nil {
+		return nil, ErrNoRegisterer
+	}
+	c := &cache[T]{
+		index:      make(map[string]*item[T]),
+		items:      make([]*item[T], 0, capacity),
+		sorted:     true,
+		capacity:   capacity,
+		metrics:    newCacheMetrics(opt.registerer, opt.extraLabels...),
+		labelsFunc: opt.labelsFunc,
+		janitor: &janitor[T]{
+			interval: opt.interval,
+			stop:     make(chan bool),
+		},
+	}
+
+	C := &Cache[T]{cache: c, keyFunc: keyFunc}
+
+	if opt.interval > 0 {
+		go c.janitor.run(c)
+	}
+
+	return C, nil
+}
+
+// Close closes the cache. It also stops the expiration eviction process.
+func (c *Cache[T]) Close() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.closed {
+		return ErrClosed
+	}
+	c.janitor.stop <- true
+	c.closed = true
+	return nil
+}
+
+// Add an item to the cache, existing index will not be overwritten.
+// To overwrite existing index, use Update.
+// If the cache is full, Add will return an error.
+func (c *Cache[T]) Add(object T) error {
+	key, err := c.keyFunc(object)
+	if err != nil {
+		return KeyError{object, err}
```

Comment:
if the keyFunc is wrong, many actions will result in error due to user provided parameter. This might have a consequence of too many metrics. Also this error is not on the cache side, so I wouldn't record it.